crates/ra_ide/src/runnables.rs

```diff
@@ -93,12 +93,23 @@ fn runnable_fn(sema: &Semantics<RootDatabase>, fn_def: ast::FnDef) -> Option<Run
 ///
 /// It may produce false positives, for example, `#[wasm_bindgen_test]` requires a different command to run the test,
 /// but it's better than not to have the runnables for the tests at all.
-fn has_test_related_attribute(fn_def: &ast::FnDef) -> bool {
-    fn_def
+fn has_test_related_attribute(fn_def: &ast::FnDef) -> Option<bool> {
+    let test = fn_def
         .attrs()
         .filter_map(|attr| attr.path())
         .map(|path| path.syntax().to_string().to_lowercase())
-        .any(|attribute_text| attribute_text.contains("test"))
+        .any(|attribute_text| attribute_text.contains("test"));
+    if test {
+        Some(
+            fn_def
```

Comment:
there should be a better way here to avoid code duplication

Reply:
I wouldn't worry about duplication here: it happens inside the function, so it's pretty benign and can be always refactored later. What I'd worry about more is the name/signature of the function:

`fn has_test_related_attribute(fn_def: &ast::FnDef) -> Option<bool> {`

"Has" function returning non-bool feels weird, and, just looking at the signature, I'd expect semantics of "definitely a test / not a test / failed to determine if a test". So, if I were to refactor this code, I'd start with doing somethig like:

```rust
struct TestAttr {
  ignore: bool,
}

impl TestAttr {
    fn from_fn(fn_def: &ast::FnDef) -> Option<TestAttr>
}
```

But than, again, this is a private function, so it doesn't matter much in the grand scheme of things if the API is less than ideal. 

Answering the direct question -- I don't think we can meaningfully remove duplication here, as the first part quite delibiratelly makes use of approximate maching. However for the second part I think we can make use of `ast::Attr::as_simple_name` (defined in extensions.rs)

Reply:
Cloning the iterator might work, this is first what comes to mind.
```
  let mut lowercased_attrs = fn_def
        .attrs()
        .filter_map(|attr| attr.path())
        .map(|path| path.syntax().to_string());
 
  lowercased_attrs
        .clone()
        .find(|attribute_text| attribute_text.contains("test"))
        .map(|_| lowercased_attrs.any(|attr| attr == "ignore"))
```

Reply:
@Veetaha thanks for the suggestion, but i implemented the suggestion to keep the name, although i could not use `as_simple_name`:

```
error[E0599]: no method named `as_simple_name` found for struct `ra_syntax::ast::generated::nodes::Attr` in the current scope
   --> crates/ra_ide/src/runnables.rs:100:37
    |
100 |             .filter_map(|attr| attr.as_simple_name())
    |                                     ^^^^^^^^^^^^^^ help: there is a method with a similar name: `as_simple_call`
```

Reply:
just found the method is called `simple_name`